### ISSUE TITLE
False positive unsupported-membership-test with Optional[str]

### ISSUE THREAD

**Issue by HWiese1980** [pylint-dev/pylint#8643]
### Bug description

I have a code snippet like this:

```python
@dataclass
class Foo:
  bar: Optional[str]

def example():
  foo = Foo("test")
  if not foo.bar or "es" not in foo.bar:
    ...
```

pylint criticizes unsupported-membership-test on the `if`-line because `foo.bar` does not support membership test. That is essentially correct if foo.bar was None. But the `is not None or` makes sure it cannot be None when the membership test is executed. Pylint does not seem to grasp this.

### Command used

```shell
Using the pre-commit hook on the entire project
```

### Pylint output

```shell
************* Module foofoo
foofoo.py:123:456: E1135: Value 'foo.bar' doesn't support membership test (unsupported-membership-test)
```

### Expected behavior

It should recognize that the membership test is fine here

### Pylint version

```shell
pylint 3.0.0a6
astroid 2.15.4
Python 3.10.8 (main, Dec  3 2022, 10:48:30) [Clang 14.0.0 (clang-1400.0.29.202)]
```

### OS / Environment

MacOS Ventura, iTerm2, oh-my-zsh, IntelliJ IDEA

### Original code from the issue including the typo

```python
@dataclass
class Foo:
  bar: Optional[str]

def example():
  foo = Foo("test")
  if not foo.bar or || "es" not in foo.bar:
    ...
```

**Comment by mbyrnepr2** [comment]
Thanks @HWiese1980. I couldn't reproduce this (note some changes below, to your code, to get it running:
```python
(venv310) Marks-MacBook-Air-2:programming markbyrne$ cat x.py 
from dataclasses import dataclass
from typing import Optional


@dataclass
class Foo:
    bar: Optional[str]

def example():
    foo = Foo("test")
    if not foo.bar or "es" not in foo.bar:
        ...
(venv310) Marks-MacBook-Air-2:programming markbyrne$ cat x.py 
from dataclasses import dataclass
from typing import Optional


@dataclass
class Foo:
    bar: Optional[str]

def example():
    foo = Foo("test")
    if not foo.bar or "es" not in foo.bar:
        ...
(venv310) Marks-MacBook-Air-2:programming markbyrne$ pylint --version
pylint 3.0.0b1
astroid 2.16.0dev0
Python 3.10.4
```

**Comment by HWiese1980** [comment]
Uh, sorry, I just saw that I had a typo in that code above. It's only supposed to serve as an MWE, written off the top of my head, and is not actual code that the error occurs with (the real code is under NDA unfortunately). I'll see if I can come up with an example that actually does reproduce the error.

**Comment by jacobtylerwalls** [comment]
Feel free to reopen when you can provide a reproducer. Thanks.
